src/server/pps/server/pipeline_controller.go

```diff
@@ -271,20 +263,58 @@ func (pc *pipelineController) newStep() (*pcStep, context.Context, error) {
 
 func (step *pcStep) updateRcAndState(ctx context.Context) error {
 	// Bring 'pipeline' into the correct state by taking appropriate action
+	if step.pipelineInfo.State == pps.PipelineState_PIPELINE_FAILURE {
+		// pipeline fails if it encounters an unrecoverable error
+		if err := step.finishPipelineOutputCommits(ctx); err != nil {
+			return err
+		}
+		// deletePipelineResources calls cancelMonitor() and cancelCrashingMonitor()
+		// in addition to deleting the RC, so those calls aren't necessary here.
+		if err := step.pc.deletePipelineResources(); err != nil {
+			// retry, but the pipeline has already failed
+			return stepError{
+				error: errors.Wrap(err, "error deleting resources for failing pipeline"),
+				retry: true,
+			}
+		}
+		return nil
+	}
+
+	if step.rc != nil && !step.rcIsFresh() {
+		// old RC is not down yet
+		return step.restartPipeline(ctx, "stale RC") // step() will be called again after collection write
+	}
+
+	if step.pipelineInfo.State == pps.PipelineState_PIPELINE_PAUSED {
```

Comment:
since there's not really a repetition/priority issue, this block could be left in the switch, just adding `&& state != PAUSED` to the condition below this